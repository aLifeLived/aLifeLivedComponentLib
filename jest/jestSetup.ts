/* eslint-disable no-undef */
jest.useFakeTimers({ legacyFakeTimers: true }); // Stops reference error from appearing
