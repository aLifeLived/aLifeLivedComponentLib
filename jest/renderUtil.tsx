import React from 'react';
import { ThemeProvider } from '../src/themes/ThemeProvider';
import { render as rtlRender } from '@testing-library/react-native';
import { library } from '@fortawesome/fontawesome-svg-core';
import { iconList } from '../src/themes/systemIcon/IconList';

library.add(iconList);

function render(ui: any, { ...renderOptions } = {}) {
  function Wrapper({ children }: { children: any }) {
    return <ThemeProvider>{children}</ThemeProvider>;
  }

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything for convenience
export * from '@testing-library/react-native';

// override render method
export { render };
