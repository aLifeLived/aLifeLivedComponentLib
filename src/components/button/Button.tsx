import { backgroundColor } from '@shopify/restyle';
import React from 'react';
import { ActivityIndicator } from 'react-native';
import {
  StyledOpacityProps,
  SystemButton,
} from '../../themes/systemButton/SystemButton';
import { SystemSpinner } from '../../themes/systemSpinner/SystemSpinner';
import { Text } from '../../themes/text/Text';

interface IButtonProps {
  title: string;
  onPress: () => void;
  isDisabled?: boolean;
  variant?: 'cta';
  color?: string;
  isLoading?: boolean;
}

export const Button: React.FC<IButtonProps & StyledOpacityProps> = ({
  title,
  onPress,
  isDisabled,
  color,
  isLoading,
  ...rest
}) => {
  return (
    <SystemButton
      backgroundColor={isDisabled ? 'contrastLow' : 'primary'}
      onPress={() => onPress()}
      padding="m"
      borderRadius="medium"
      width="90%"
      disabled={isDisabled || isLoading}
      {...rest}
    >
      {isLoading ? (
        <SystemSpinner
          size="small"
          color="white"
          testID="QBYQCQ-button-spinner"
          animating
        />
      ) : (
        <Text
          variant="default"
          fontWeight="600"
          textAlign="center"
          color={color}
        >
          {title}
        </Text>
      )}
    </SystemButton>
  );
};
