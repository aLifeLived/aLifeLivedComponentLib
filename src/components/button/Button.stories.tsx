import React from 'react';

import { storiesOf } from '@storybook/react-native';

// Component
import { Button } from './Button';
import { action } from '@storybook/addon-actions';
import { CenterView } from '../centerView/CenterView';
import { Box } from '../../themes/box/Box';

storiesOf('Button', module)
  .addDecorator(Story => <CenterView>{Story()}</CenterView>)
  .add('Default button', () => (
    <Box width="100%" alignItems="center">
      <Button
        title="Test button that has a long title"
        onPress={action('Button on press fired')}
        isLoading={false}
        isDisabled={false}
        margin="m"
        backgroundColor="primaryRed"
      />

      <Button
        title="Loading button"
        onPress={action('Button on press')}
        isLoading={true}
        isDisabled={false}
        margin="m"
        width="90%"
      />

      <Button
        title="Disabled button"
        onPress={action('Button on press')}
        isLoading={false}
        isDisabled={true}
        margin="m"
        width="90%"
      />
    </Box>
  ));
