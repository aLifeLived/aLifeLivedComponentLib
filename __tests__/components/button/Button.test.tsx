import React from 'react';
import { render, fireEvent } from './../../../jest/renderUtil';

// Component
import { Button } from '../../../src/components/button/Button';

const mockOnPress = jest.fn();

const setup = ({ isDisabled = false, isLoading = false }) => {
  // Screen
  const utils = render(
    <Button
      title="Test button"
      onPress={mockOnPress}
      isDisabled={isDisabled}
      isLoading={isLoading}
    />
  );

  return {
    ...utils,
  };
};

describe('components >> button ', () => {
  describe('Enabled button', () => {
    test('Should render basic button correctly', () => {
      const { toJSON } = setup({ isDisabled: false, isLoading: false });
      expect(toJSON()).toMatchSnapshot();
    });

    test('Should render title', () => {
      const { getByText } = setup({ isDisabled: false, isLoading: false });
      expect(getByText('TEST BUTTON')).not.toBeNull();
    });

    test('Should fire on press', () => {
      const { getByText } = setup({ isDisabled: false, isLoading: false });
      fireEvent.press(getByText('TEST BUTTON'));
      expect(mockOnPress).toBeCalledTimes(1);
    });
  });

  describe('Disabled button', () => {
    beforeAll(() => mockOnPress.mockClear());
    test('Should render basic button correctly', () => {
      const { toJSON } = setup({ isDisabled: true, isLoading: false });
      expect(toJSON()).toMatchSnapshot();
    });

    test('Should find title', () => {
      const { getByText } = setup({ isDisabled: true, isLoading: false });
      expect(getByText('TEST BUTTON')).toBeDefined();
    });

    test('Should not be able to fire on press action', () => {
      const { getByText } = setup({ isDisabled: true, isLoading: false });
      fireEvent.press(getByText('TEST BUTTON'));
      expect(mockOnPress).toBeCalledTimes(0);
    });
  });

  describe('Loading button', () => {
    beforeAll(() => mockOnPress.mockClear());
    test('Should render button in loading state', () => {
      const { toJSON } = setup({ isDisabled: false, isLoading: true });
      expect(toJSON()).toMatchSnapshot();
    });

    test('Should NOT render title', () => {
      const { queryByText } = setup({ isDisabled: false, isLoading: true });
      expect(queryByText('TEST BUTTON')).toBeNull();
    });

    test('Should NOT fire on press', () => {
      const { getByTestId } = setup({ isDisabled: false, isLoading: true });
      fireEvent.press(getByTestId('QBYQCQ-button-spinner'));
      expect(mockOnPress).toBeCalledTimes(0);
    });
  });
});
