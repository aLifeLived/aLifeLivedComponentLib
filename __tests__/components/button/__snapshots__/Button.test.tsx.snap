// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components >> button  Disabled button Should render basic button correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#54575a",
      "borderRadius": 8,
      "opacity": 1,
      "padding": 16,
      "width": "90%",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "display": "flex",
          "fontSize": 16,
          "fontWeight": "600",
          "lineHeight": 20,
          "textAlign": "center",
        },
      ]
    }
  >
    Test button
  </Text>
</View>
`;

exports[`components >> button  Enabled button Should render basic button correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#0FCFA6",
      "borderRadius": 8,
      "opacity": 1,
      "padding": 16,
      "width": "90%",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "display": "flex",
          "fontSize": 16,
          "fontWeight": "600",
          "lineHeight": 20,
          "textAlign": "center",
        },
      ]
    }
  >
    Test button
  </Text>
</View>
`;

exports[`components >> button  Loading button Should render button in loading state 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#0FCFA6",
      "borderRadius": 8,
      "opacity": 1,
      "padding": 16,
      "width": "90%",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
    style={
      Array [
        Object {
          "color": "#252424",
        },
      ]
    }
    testID="QBYQCQ-button-spinner"
  />
</View>
`;
